File: scribe/watcher.py
"""Filesystem watcher built on watchdog.

OBS writes recordings progressively, so a file is only enqueued once its
size stays identical across ``stability_checks`` consecutive probes spaced
``stability_interval_seconds`` apart. Events land in a queue consumed by a
daemon worker thread that runs the pipeline sequentially. On startup the
inbox is scanned for existing, not-yet-processed videos, which are enqueued
too.
"""

from __future__ import annotations

import logging
import queue
import threading
import time
from pathlib import Path
from typing import Callable

from watchdog.events import FileSystemEventHandler
from watchdog.observers import Observer

from .config import Config

log = logging.getLogger(__name__)


class _RecordingHandler(FileSystemEventHandler):
    """Forwards created/moved video files to the stability-checking enqueue callback."""

    def __init__(self, config: Config, on_candidate: Callable[[Path], None]) -> None:
        super().__init__()
        self._config = config
        self._on_candidate = on_candidate

    def _maybe_enqueue(self, path_str: str) -> None:
        path = Path(path_str)
        if self._config.is_video(path):
            self._on_candidate(path)

    def on_created(self, event) -> None:
        if not event.is_directory:
            self._maybe_enqueue(event.src_path)

    def on_moved(self, event) -> None:
        if not event.is_directory:
            self._maybe_enqueue(event.dest_path)


def wait_until_stable(path: Path, checks: int, interval: float) -> bool:
    """True when the file size is identical across ``checks`` probes.

    ``checks`` identical consecutive reads (each ``interval`` seconds after
    the previous one) mean OBS finished writing. Returns False if the file
    disappears mid-wait.
    """
    try:
        last_size = path.stat().st_size
    except OSError:
        return False
    stable = 0
    while stable < checks:
        time.sleep(interval)
        try:
            size = path.stat().st_size
        except OSError:
            return False
        if size == last_size:
            stable += 1
        else:
            stable = 0
            last_size = size
    return True


class Watcher:
    def __init__(self, config: Config, process_fn: Callable[[Path], None]) -> None:
        self._config = config
        self._process_fn = process_fn
        self._queue: queue.Queue[Path] = queue.Queue()
        self._pending: set[str] = set()
        self._pending_lock = threading.Lock()
        self._observer = Observer()

    # ------------------------------------------------------------- enqueueing

    def _submit(self, path: Path) -> None:
        """Enqueue a candidate after its size stabilizes (deduplicated)."""
        with self._pending_lock:
            if str(path) in self._pending:
                return
            self._pending.add(str(path))
        try:
            log.info("New recording detected: %s (waiting for it to finish writing)", path.name)
            if wait_until_stable(path, self._config.stability_checks,
                                 self._config.stability_interval_seconds):
                log.info("File stable, enqueueing: %s", path.name)
                self._queue.put(path)
            else:
                log.warning("File vanished before stabilizing: %s", path.name)
        finally:
            with self._pending_lock:
                self._pending.discard(str(path))

    def submit_async(self, path: Path) -> None:
        """Stability check runs in its own thread so watchdog events never block."""
        threading.Thread(target=self._submit, args=(path,), daemon=True).start()

    def scan_existing(self) -> None:
        """Enqueue unprocessed videos already sitting in the inbox."""
        if not self._config.inbox.is_dir():
            return
        for path in sorted(self._config.inbox.iterdir()):
            if path.is_file() and self._config.is_video(path):
                self.submit_async(path)

    # ---------------------------------------------------------------- workers

    def _worker(self) -> None:
        while True:
            path = self._queue.get()
            try:
                self._process_fn(path)
            except Exception:
                log.exception("Failed to process %s", path)
            finally:
                self._queue.task_done()

    # ------------------------------------------------------------------- run

    def run(self) -> None:
        self._config.inbox.mkdir(parents=True, exist_ok=True)
        handler = _RecordingHandler(self._config, self.submit_async)
        self._observer.schedule(handler, str(self._config.inbox), recursive=False)

        threading.Thread(target=self._worker, daemon=True).start()

        log.info("Watching %s for %s files ...",
                 self._config.inbox, ", ".join(self._config.video_extensions))
        self.scan_existing()

        self._observer.start()
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            log.info("Shutting down ...")
        finally:
            self._observer.stop()
            self._observer.join()

File: internal/watcher/watcher.go
// Package watcher watches the inbox folder for new video recordings.
//
// OBS writes recordings progressively, so a file is only enqueued once its
// size stays identical across Config.StabilityChecks consecutive probes
// spaced Config.StabilityIntervalSeconds apart. Stable files land in a queue
// consumed by a single worker goroutine that runs processFn sequentially.
// On startup the inbox is scanned for existing videos, which take the same
// path.
package watcher

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"

	"github.com/fernando143/patro/internal/config"
	"github.com/fernando143/patro/internal/logging"
	"github.com/fernando143/patro/internal/status"
)

// WaitUntilStable reports whether the file size stayed identical across
// checks consecutive probes spaced interval apart. checks identical
// consecutive reads (each interval after the previous one) mean OBS finished
// writing. It returns false if the file disappears mid-wait.
func WaitUntilStable(path string, checks int, interval time.Duration) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	lastSize := info.Size()
	stable := 0
	for stable < checks {
		time.Sleep(interval)
		info, err := os.Stat(path)
		if err != nil {
			return false
		}
		if size := info.Size(); size == lastSize {
			stable++
		} else {
			stable = 0
			lastSize = size
		}
	}
	return true
}

// Watcher watches the inbox directory and feeds stable video files to
// processFn one at a time.
type Watcher struct {
	cfg       *config.Config
	processFn func(path string)

	// Tracker publishes queue/processing state for the dashboard. It is
	// optional and safe to leave nil (all its methods are nil-safe).
	Tracker *status.Tracker

	// interval overrides the stability-probe interval when > 0; Config only
	// stores whole seconds, which is too coarse for tests.
	interval time.Duration

	queue      chan string
	pendingMu  sync.Mutex
	pending    map[string]struct{}
	submitters sync.WaitGroup
	workerWG   sync.WaitGroup
}

// New returns a Watcher that calls processFn for each stable video file
// appearing in cfg.Inbox.
func New(cfg *config.Config, processFn func(path string)) *Watcher {
	return &Watcher{
		cfg:       cfg,
		processFn: processFn,
		queue:     make(chan string),
		pending:   map[string]struct{}{},
	}
}

// Run watches cfg.Inbox until ctx is cancelled.
//
// The inbox is created if missing. Per-file failures (vanished files,
// processFn errors or panics) are logged, never returned; only setup
// failures produce a non-nil error. On cancellation the fsnotify watch is
// removed and closed, then Run waits for in-flight stability checks and
// queued files to finish before returning nil, so a blocked processFn
// delays shutdown.
func (w *Watcher) Run(ctx context.Context) error {
	if err := os.MkdirAll(w.cfg.Inbox, 0o755); err != nil {
		return fmt.Errorf("watcher: cannot create inbox %s: %w", w.cfg.Inbox, err)
	}
	fsw, err := fsnotify.NewWatcher()
	if err != nil {
		return fmt.Errorf("watcher: %w", err)
	}
	if err := fsw.Add(w.cfg.Inbox); err != nil {
		fsw.Close()
		return fmt.Errorf("watcher: cannot watch %s: %w", w.cfg.Inbox, err)
	}

	w.workerWG.Add(1)
	go w.worker()

	logging.Infof("Watching %s for %s files ...", w.cfg.Inbox, strings.Join(w.cfg.VideoExtensions, ", "))
	w.scanExisting()

loop:
	for {
		select {
		case <-ctx.Done():
			break loop
		case event, ok := <-fsw.Events:
			if !ok {
				break loop
			}
			w.handleEvent(event)
		case err, ok := <-fsw.Errors:
			if !ok {
				break loop
			}
			logging.Errorf("Filesystem watcher error: %v", err)
		}
	}

	logging.Infof("Shutting down ...")
	fsw.Close()
	w.stop()
	return nil
}

// stop waits for in-flight stability checks (the worker keeps draining the
// queue meanwhile), then closes the queue and waits for the worker to finish
// the remaining files.
func (w *Watcher) stop() {
	w.submitters.Wait()
	close(w.queue)
	w.workerWG.Wait()
}

// handleEvent reacts to a new inbox entry. A file moved INTO the directory
// surfaces as Create, so Create alone covers Python's on_created + on_moved.
func (w *Watcher) handleEvent(event fsnotify.Event) {
	if !event.Has(fsnotify.Create) {
		return
	}
	if info, err := os.Stat(event.Name); err == nil && info.IsDir() {
		return
	}
	if w.cfg.IsVideo(event.Name) {
		w.submitAsync(event.Name)
	}
}

// scanExisting submits video files already sitting in the inbox. os.ReadDir
// returns entries sorted by name, matching Python's sorted() scan.
func (w *Watcher) scanExisting() {
	entries, err := os.ReadDir(w.cfg.Inbox)
	if err != nil {
		logging.Errorf("Cannot scan inbox %s: %v", w.cfg.Inbox, err)
		return
	}
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		path := filepath.Join(w.cfg.Inbox, entry.Name())
		if w.cfg.IsVideo(path) {
			w.submitAsync(path)
		}
	}
}

// submitAsync runs the stability check in its own goroutine so fsnotify
// events never block; a path with a check already in flight is dropped.
func (w *Watcher) submitAsync(path string) {
	w.pendingMu.Lock()
	if _, dup := w.pending[path]; dup {
		w.pendingMu.Unlock()
		return
	}
	w.pending[path] = struct{}{}
	w.pendingMu.Unlock()

	w.submitters.Add(1)
	go func() {
		defer w.submitters.Done()
		defer func() {
			w.pendingMu.Lock()
			delete(w.pending, path)
			w.pendingMu.Unlock()
		}()

		name := filepath.Base(path)
		logging.Infof("New recording detected: %s (waiting for it to finish writing)", name)
		if WaitUntilStable(path, w.cfg.StabilityChecks, w.stabilityInterval()) {
			logging.Infof("File stable, enqueueing: %s", name)
			w.Tracker.Enqueue(path)
			w.queue <- path
		} else {
			logging.Warnf("File vanished before stabilizing: %s", name)
		}
	}()
}

// worker consumes stable files sequentially, recovering a panic per file
// (parity with Python's try/except around process_fn).
func (w *Watcher) worker() {
	defer w.workerWG.Done()
	for path := range w.queue {
		w.Tracker.Dequeue(path)
		func() {
			defer func() {
				if r := recover(); r != nil {
					logging.Errorf("Failed to process %s: %v", path, r)
					w.Tracker.Fail(path, fmt.Sprintf("panic: %v", r))
				}
			}()
			w.processFn(path)
		}()
	}
}

// stabilityInterval returns the probe interval: the override when set,
// otherwise the configured whole seconds.
func (w *Watcher) stabilityInterval() time.Duration {
	if w.interval > 0 {
		return w.interval
	}
	return time.Duration(w.cfg.StabilityIntervalSeconds) * time.Second
}
